File: element-recursion/josephus.rb
class Josephus
    def initialize(count,nth)
        @count = count
        @nth = nth
    end
    def shout()
        shout_recursive((1..@count).to_a, @nth, 0)
    end
    def shout_recursive(people,nth,counter)
        if people.size > 1
            if counter == 0
                people.shift
                counter = counter + 1
            else
                if counter == nth - 1
                    counter = 0
                else
                    counter = counter + 1
                end
                people.rotate!
            end
            return shout_recursive(people,nth,counter)
        else
            return people[0]
        end
    end
end

j = Josephus.new(40,3)
puts j.shout()
TIMES=10
ITER=1000000
TIMES.times { |i|
    start = Time.now
    ITER.times { |k|
        j.shout()
    }
    ends = Time.now
    puts ((ends - start) * 1000000 / ITER).to_s()
}

File: element-recursion/josephus.erl
-module(josephus).
-export([benchmark/0]).
-compile(native).

shout(Count,Nth) ->
    shout(lists:seq(1,Count),[],Nth,1).

shout([Head | []], [], _, _) -> Head;
shout([], Survivors, Nth, Counter) -> 
    %% io:format("Reversing~n",[]),
    shout(lists:reverse(Survivors),[], Nth, Counter);
shout([_ | Tail], Survivors, Nth, 1) ->
    %% io:format("~w dies~n",[Head]),
    shout(Tail, Survivors, Nth, 2);
shout([Head | Tail], Survivors, Nth, Nth) ->
    %% io:format("~w survives~n",[Head]),
    shout(Tail, [Head | Survivors], Nth, 1);
shout([Head | Tail], Survivors, Nth, Counter) ->
    %% io:format("~w survives~n",[Head]),
    shout(Tail, [Head | Survivors], Nth, Counter + 1).

run_iterations(0) -> ok;
run_iterations(Iterations) ->
    shout(40,3),
    run_iterations(Iterations - 1).

run_times(_,0) -> ok;
run_times(Iterations, Times) ->
    Start = os:timestamp(),
    run_iterations(Iterations),
    End = os:timestamp(),
    io:format("~w~n",[timer:now_diff(End,Start) / Iterations]),
    run_times(Iterations,Times-1).

benchmark() ->
    Iter = 1000000,
    io:format("(element recursive)~n",[]),
    run_times(Iter,10).
